src/doc/trpl/concurrency.md

```diff
@@ -0,0 +1,371 @@
+incorrectly also helps rule out data races, one of the worst kinds of
+concurrency bugs.
+
+As an example, here is a Rust program that would have a data race in many
+languages. It will not compile:
+
+```ignore
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+
+fn main() {
+    let mut data = vec![1u32, 2, 3];
+
+    for i in 0 .. 2 {
+        Thread::spawn(move || {
+            data[i] += 1;
+        });
+    }
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+This gives us an error:
+
+```text
+12:17 error: capture of moved value: `data`
+        data[i] += 1;
+        ^~~~
+```
+
+In this case, we know that our code _should_ be safe, but Rust isn't sure. And
+it's actually not safe: if we had a reference to `data` in each thread, and the
+thread takes ownership of the reference, we have three owners! That's bad. We
+can fix this by using the `Arc<T>` type, which provides an atmoic reference
+count. The 'atomic' part means that it's safe to share across threads.
+
+`Arc<T>` assumes one more property about its contents to ensure that it is safe
+to share across threads: it assumes its contents are immutable. But in our
+case, we want to be able to mutate the value. We need a type that can ensure
+only one person at a time can mutate what's inside. For that, we can use the
+`Mutex<T>` type. Here's the second version of our code. It still doesn't work,
+but for a different reason:
+
+```ignore
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+use std::sync::Mutex;
+
+fn main() {
+    let mut data = Mutex::new(vec![1u32, 2, 3]);
+
+    for i in 0 .. 2 {
+        let data = data.lock().unwrap();
+        Thread::spawn(move || {
+            data[i] += 1;
+        });
+    }
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+Here's the error:
+
+```text
+error: the trait `core::marker::Send` is not implemented for the type `core::marker::NoSend`
+```
+
+You see, [`Mutex`](std/sync/struct.Mutex.html) has a
+[`lock`](http://doc.rust-lang.org/nightly/std/sync/struct.Mutex.html#method.lock)
+method which has this siganture:
+
+```ignore
+fn lock(&self) -> LockResult<MutexGuard<T>>
+```
+
+If we [look at the code for MutexGuard](https://github.com/rust-lang/rust/blob/896cb36ecab3eaeb7f101087e030e43771eca5ca/src/libstd/sync/mutex.rs#L163-L170), we'll see
+this:
+
+```ignore
+__marker: marker::NoSend,
+```
+
+Because our guard is `NoSend`, it's not `Send`. Which means we can't actually
+transfer the guard across task boundaries, which gives us our error.
+
+The [`Arc<T>`](std/sync/struct.Arc.html) type is the answer. An `Arc<T>` allows
+us to reference count some data using atomic operations, which will allow our
+code to work. Here's the working version:
+
+```
+use std::sync::{Arc, Mutex};
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+
+fn main() {
+    let data = Arc::new(Mutex::new(0u32));
```

Comment:
+1 for consistently with earlier examples.

I was also confused as to why, when the doc zeroes in on `marker::NoSend` as the big problem, the next sentence presents `Arc` as the solution.  It seems to me like the proximal solution was to move `data.lock().unwrap()` inside the thread, so that the guard is not sent into the closure.  I guess `Arc` is somehow necessary to move that line inside the closure, but the mechanism isn't clear.

I think the pedagogical approach of showing a series of failed attempts is very useful.  Perhaps this section could benefit from adding yet another failed attempt to show the problem with moving `data.lock.unwrap()` inside the closure without `Arc`.  After that we would see the final `Arc` example.  That structure would be more precise about the problem `Arc` solves.